test-series/4/tools/sim_prep.py

```diff
@@ -0,0 +1,626 @@
+    
+    After downselection and rephasing, this function overwrites the 
+    simulation LST and time arrays to match the reference LSTs and times.
+    
+    Parameters
+    ----------
+    sim_uvd : :class:`pyuvdata.UVData`
+        :class:`pyuvdata.UVData` object containing the simulation data.
+        
+    ref_uvd : :class:`pyuvdata.UVData`
+        :class:`pyuvdata.UVData` object containing the reference data. 
+        Only the metadata for this object needs to be read.
+
+    Returns
+    -------
+    new_sim_uvd : :class:`pyuvdata.UVData`
+        :class:`pyuvdata.UVData` object containing the modified 
+        simulation data. The modified data has been rephased to match 
+        the reference LSTs and has had its time and LST arrays adjusted 
+        to contain only the times and LSTs that are present in the 
+        reference data.
+    """
+    # Convert the simulation to a HERAData object. 
+    sim_uvd = to_HERAData(copy.deepcopy(sim_uvd))
+    hd_metas = sim_uvd.get_metadata_dict()
+
+    # Load in useful metadata.
+    sim_lsts = hd_metas['lsts']
+    sim_times = hd_metas['times']
+    ref_lsts = np.unique(ref_uvd.lst_array)
+    ref_times = np.unique(ref_uvd.time_array)
+    sim_to_ref_time_map = get_d2m_time_map(sim_times, sim_lsts, ref_times, ref_lsts)
+    use_times = np.array([
+        sim_time for sim_time, ref_time in sim_to_ref_time_map.items()
+        if ref_time is not None
+    ])
+    use_lsts = np.array(
+        [sim_lsts[np.argmin(np.abs(sim_times - time))] for time in use_times]
+    )
+    ref_lsts = np.array([
+        ref_lsts[np.argmin(np.abs(ref_times - sim_to_ref_time_map[sim_time]))] 
+        for sim_time in use_times
+    ])
+    ref_times = np.array([sim_to_ref_time_map[sim_time] for sim_time in use_times])
+
+    # Downselect in time and load data.
+    sim_uvd.select(times=use_times)
+    data, _, _ = sim_uvd.build_datacontainers()
+
+    # Build the antpair -> ENU baseline vector dictionary.
+    antpos, ants = sim_uvd.get_ENU_antpos()
+    bls = {bl : None for bl in data.bls()}
+    for bl in bls:
+        ai, aj = bl[:2]
+        i, j = ants.tolist().index(ai), ants.tolist().index(aj)
+        bls[bl] = antpos[j] - antpos[i]
+
+    # Rephase and update the data.
+    new_data, new_flags = rephase_vis(
+        data, use_lsts, ref_lsts, bls, hd_metas['freqs']
+    )
+    sim_uvd.update(data=new_data, flags=new_flags)
+    new_sim_lsts = np.zeros_like(sim_uvd.lst_array)
+    new_sim_times = np.zeros_like(sim_uvd.time_array)
+    for use_time, ref_time, ref_lst in zip(use_times, ref_times, ref_lsts):
+        blt_slice = np.argwhere(sim_uvd.time_array == use_time)
+        new_sim_lsts[blt_slice] = ref_lst
+        new_sim_times[blt_slice] = ref_time
+    sim_uvd.time_array = new_sim_times
+    sim_uvd.lst_array = new_sim_lsts
+
+    # Ensure that we return a UVData object so that chunk_sim_and_save doesn't break.
+    return_uvd = UVData()
```

Comment:
That's what I would've thought as well, but that doesn't seem to be the case. Another issue I had with it was that the `to_HERAData` function actually changes the class of the object that's passed to it, but I wasn't sure whether to raise an issue on the hera_cal github since it's in the legacy code section of the module.

Reply:
wow... didn't even know that was possible! Really, `HERAData` should have a `.from_uvdata` classmethod.